MovingBlocks/DestinationSol
# Fix distribution creation and add modules to exported distribution

# Description
This pull request fixes the `gradlew dist` and `gradlew distZip` commands. It also adds support for including all the modules from the current workspace in the distribution,

The modules will be compiled into jars and copied into the `modules` directory when the distribution is created. Both the code and the assets should be contained within a single jar for each module, named after the module and its version.

Directory-based modules, such as those used for development, should still work when copied into the output modules directory. Zip archives do not work currently.

# Testing
## Testing distribution creation
- Run the command `gradlew dist`.
- A folder, named `app`, should be created within the `desktop/build/distributions/` directory. Enter this directory.
- Execute the relevant launcher (`sol.exe` on Windows, `solOSX.sh` on MacOS and `sol[64].sh` on Linux). The game should start and should be able to be played normally, without any additional crashes.

## Testing distribution zip creation
- Run the command `gradlew distZip`
- A zip archive should be created, called `DestinationSol.zip`, within the `desktop/build/distributions` directory.
- Extract `DestinationSol.zip` into a directory and enter it,
- Execute the relevant launcher (`sol.exe` on Windows, `solOSX.sh` on MacOS and `sol[64].sh` on Linux). The game should start and should be able to be played normally, without any additional crashes.

## Testing modules
### Testing code-only modules
- Run the command `groovyw module get codeModuleTest -remote BenjaminAmos` to get my codeModuleTest module.
- Follow the steps contained within the "Testing distribution creation" section.
- Check that in the top-left corner of the screen there is some text labelled "Game Duration: ", followed by the quantity of seconds that the current game session has run for.

### Testing asset-only modules
- Follow the steps contained within the "Testing distribution creation" section.
- Check that the game does not crash when starting a new game. If the core module is not present or loaded then the game will crash.

### Testing modules containing both assets and code
- Run the command `groovyw module get warp` to get the warp module.
- Follow the steps contained within the "Testing distribution creation" section.
- Start a new game with the "Warp Scout" ship.
- Check that the game does not crash.
- Check that there is a button on the right side of the screen labelled "Research".



# Notes
- A dependency on the package `com.github.marschall:zipfilesystem-standalone` has been added to the engine module in order to fix the built-in jre distribution (see https://github.com/MovingBlocks/DestinationSol/pull/300#issuecomment-497719782 for more information).
- This has not been tested on Android.
- I have tested this on Windows and Linux (Ubuntu 19.04) but not on MacOS.
- Modules are not loaded successfully when provided as zip archives (.zip).
- On Linux, the executable permissions are broken, so you will need to add execute rights to the following files:
  - `lwjreLinux/bin/java`
  - `sol.sh`
  - `sol64.sh`
- If merged, this can be considered to supersede #300.
- This fixes #181.

# Credits
- To **Avalancs** for providing the intial changes that partially resolved the issue (see #181)
- To **NicholasBatesNZ** for providing additional improvements to **Avalancs**'s code (see #300)
- To **Adrijaned** for making clear that build.gradle should be ignored in modules (see https://github.com/MovingBlocks/DestinationSol/pull/300#issuecomment-497729991)
- I referenced Terasology's code for this and certain methods used may be similar, if not identical, to code in the following files:
  - [build.gradle](https://github.com/MovingBlocks/Terasology/blob/dbc32f0da97e46c447ae2293a00e5babaf92c85b/build.gradle#L145-L148)
  - [facades/PC/build.gradle](https://github.com/MovingBlocks/Terasology/blob/dbc32f0da97e46c447ae2293a00e5babaf92c85b/facades/PC/build.gradle#L225-L277)
  - [config/groovy/module.groovy](https://github.com/MovingBlocks/Terasology/blob/dbc32f0da97e46c447ae2293a00e5babaf92c85b/config/groovy/module.groovy#L46-L63)
